test/core/event_engine/test_suite/server_test.cc

```diff
@@ -12,12 +12,263 @@
+// For each data transfer, verify that data written at one end of the stream
+// equals data read at the other end of the stream.
+TEST_F(EventEngineServerTest, ServerConnectExchangeBidiDataTransferTest) {
+  grpc_core::ExecCtx ctx;
+  auto oracle_ee = this->NewOracleEventEngine();
+  std::shared_ptr<EventEngine> test_ee(this->NewEventEngine());
+  auto memory_quota = std::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  std::unique_ptr<EventEngine::Endpoint> client_endpoint;
+  std::unique_ptr<EventEngine::Endpoint> server_endpoint;
+  grpc_core::Notification client_signal;
+  grpc_core::Notification server_signal;
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint, &server_signal](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint = std::move(ep);
+        server_signal.Notify();
+      };
+
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  ChannelArgsEndpointConfig config(args);
+  auto status = test_ee->CreateListener(
+      std::move(accept_cb), [](absl::Status /*status*/) {}, config,
+      std::make_unique<grpc_core::MemoryQuota>("foo"));
+  EXPECT_TRUE(status.ok());
+
+  std::unique_ptr<Listener> listener = std::move(*status);
+  EXPECT_TRUE(listener->Bind(URIToResolvedAddress(target_addr)).ok());
+  EXPECT_TRUE(listener->Start().ok());
+
+  oracle_ee->Connect(
+      [&client_endpoint,
+       &client_signal](absl::StatusOr<std::unique_ptr<Endpoint>> status) {
+        if (!status.ok()) {
+          gpr_log(GPR_ERROR, "Connect failed: %s",
+                  status.status().ToString().c_str());
+          client_endpoint = nullptr;
+        } else {
+          client_endpoint = std::move(*status);
+        }
+        client_signal.Notify();
+      },
+      URIToResolvedAddress(target_addr), config,
+      memory_quota->CreateMemoryAllocator("conn-1"), 24h);
+
+  client_signal.WaitForNotification();
+  server_signal.WaitForNotification();
+  EXPECT_NE(client_endpoint.get(), nullptr);
+  EXPECT_NE(server_endpoint.get(), nullptr);
+
+  // Alternate message exchanges between client -- server and server --
+  // client.
+  for (int i = 0; i < kNumExchangedMessages; i++) {
+    // Send from client to server and verify data read at the server.
+    EXPECT_TRUE(SendValidatePayload(GetNextSendMessage(), client_endpoint.get(),
+                                    server_endpoint.get())
+                    .ok());
+
+    // Send from server to client and verify data read at the client.
+    EXPECT_TRUE(SendValidatePayload(GetNextSendMessage(), server_endpoint.get(),
+                                    client_endpoint.get())
+                    .ok());
+  }
+  client_endpoint.reset();
```

Comment:
Are these being reset to ensure they don't hold EventEngine refs?

Reply:
Yeah. Otherwise the subsequent WaitForSingleOwner call will not return.

Reply:
Ok, thanks for clarifying.